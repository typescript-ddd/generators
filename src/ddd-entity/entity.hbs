import { CreateEntityProps, Entity, UpdateEntityProps } from "@typescript-ddd/core";
import { {{pascal entityName}}Id } from "./{{lower entityName}}-id";

/**
 * Represents the properties required to create {{noun entityName}}.
 */
export interface Create{{pascal entityName}}Props extends CreateEntityProps {
  {{#each properties}}
  {{#if isNullable}}
  {{name}}?: {{type}},
  {{else}}
  {{name}}: {{type}},
  {{/if}}
  {{/each}}
}

/**
 * Represents the properties required to update {{noun entityName}}.
 */
export interface Update{{pascal entityName}}Props extends UpdateEntityProps<{{pascal entityName}}Id> {
  {{#each properties}}
  {{#if isNullable}}
  {{name}}?: {{type}},
  {{else}}
  {{name}}: {{type}},
  {{/if}}
  {{/each}}
}

/**
 * Represents {{noun entityName}}.
 */
export class {{pascal entityName}} extends Entity<{{pascal entityName}}Id> {
  /**
   * Initilizes a new instance of a {{natural entityName}}.
   * @param {{lbrace}}{{pascal entityName}}Id{{rbrace}} id - The unique identifier of the {{natural entityName}}.
   {{#each properties}}
   * @param {{lbrace}}{{type}}{{rbrace}} {{name}} - The {{name}} of the {{natural ../entityName}}.
   {{/each}}
   */
  private constructor(
    id: {{pascal entityName}}Id,    
    {{#each properties}}
    private readonly {{name}}{{#if isNullable}}?{{/if}}: {{type}},
    {{/each}}) {
      super(id);
    }

  {{#each properties}}
  /**
   * Gets the {{name}}.
   * @returns {{lbrace}}{{type}}{{#if isNullable}} | undefined{{/if}}{{rbrace}}
   */
   get{{pascal name}}(): {{type}}{{#if isNullable}} | undefined{{/if}} {
     return this.{{name}};
   }

  {{/each}}
  /**
   * Creates a new instance of {{noun entityName}}.
   * @param {{lbrace}}Create{{pascal entityName}}Props{{rbrace}} props - The properties to create the {{natural entityName}}.
   * @param {{lbrace}}{{pascal entityName}}Id{{rbrace}} id - Optional unique identifier of the {{natural entityName}}.
   * @returns {{lbrace}}{{pascal entityName}}{{rbrace}} - A new instance of {{noun entityName}}.
   */
  static create(props: Create{{pascal entityName}}Props, id?: {{pascal entityName}}Id): {{pascal entityName}} {
    return new {{pascal entityName}}(
      id ?? {{pascal entityName}}Id.create(),
      {{#each properties}}
      props.{{name}},
      {{/each}}
    );
  }
}