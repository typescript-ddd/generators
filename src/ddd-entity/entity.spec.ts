import { EntityTemplateData } from "./entity.types";
import { generateEntity } from "./entity.generator";

describe("Entity", () => {
  it("should render template", () => {
    // Arrange
    const data: EntityTemplateData = {
      entityName: "User",
      properties: [
        { name: "name", type: "string" },
        { name: "email", type: "string" },
      ],
    };

    // Act
    const result = generateEntity(data);

    // Assert
    expect(result).toMatchSnapshot();
  });
});
