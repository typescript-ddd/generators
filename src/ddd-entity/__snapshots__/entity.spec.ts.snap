// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Entity should render template 1`] = `
"import { CreateEntityProps, Entity, UpdateEntityProps } from "@typescript-ddd/core";
import { UserId } from "./user-id";

/**
 * Represents the properties required to create a user.
 */
export interface CreateUserProps extends CreateEntityProps {
  name: string,
  email: string,
}

/**
 * Represents the properties required to update a user.
 */
export interface UpdateUserProps extends UpdateEntityProps<UserId> {
  name: string,
  email: string,
}

/**
 * Represents a user.
 */
export class User extends Entity<UserId> {
  /**
   * Initilizes a new instance of a user.
   * @param {UserId} id - The unique identifier of the user.
   * @param {string} name - The name of the user.
   * @param {string} email - The email of the user.
   */
  private constructor(
    id: UserId,    
    private readonly name: string,
        private readonly email: string,
    ) {
      super(id);
    }

  /**
   * Gets the name.
   * @returns {string}
   */
   getName(): string {
     return this.name;
   }

  /**
   * Gets the email.
   * @returns {string}
   */
   getEmail(): string {
     return this.email;
   }

  /**
   * Creates a new instance of a user.
   * @param {CreateUserProps} props - The properties to create the user.
   * @param {UserId} id - Optional unique identifier of the user.
   * @returns {User} - A new instance of a user.
   */
  static create(props: CreateUserProps, id?: UserId): User {
    return new User(
      id ?? UserId.create(),
      props.name,
      props.email,
    );
  }
}"
`;
